lib/value/src/kind/accessor.rs

```diff
@@ -0,0 +1,543 @@
+use std::collections::BTreeMap;
+
+use super::{Collection, Kind};
+use lookup::{Lookup, Segment};
+
+impl Kind {
+    /// Find the [`Kind`] at the given path.
+    ///
+    /// If the path points to root, then `self` is returned, otherwise `None` is returned if `Kind`
+    /// isn't an object or array, or if the path points to a non-existing field/index in the
+    /// object/array.
+    ///
+    /// Negative indexing always returns `None`, as the type system doesn't know the exact size of
+    /// an array, and thus cannot count backward.
+    #[must_use]
+    pub fn find_at_path(&self, path: &Lookup<'_>) -> Option<&Kind> {
+        if path.is_root() {
+            return Some(self);
+        }
+
+        let mut kind = self;
+        for segment in path.iter() {
+            kind = match segment {
+                // Try finding the field in the existing object.
+                Segment::Field(field) => kind
+                    .object
+                    .as_ref()
+                    .and_then(|collection| collection.known().get(&(field.into())))?,
+
+                // Try finding one of the fields in the existing object.
+                Segment::Coalesce(fields) => kind.object.as_ref().and_then(|collection| {
+                    let field = fields
+                        .iter()
+                        .find(|field| collection.known().contains_key(&((*field).into())))?;
+
+                    collection.known().get(&(field.into()))
```

Comment:
I'm not sure if this is correct.

Say we have an object `{thing: int | null, thung: float}` and we do `find_at_path(".(thing | thung)")`. 

I believe this should return `int | float`, since the resulting type at runtime of fetching that path would either be an `int` or a `float`, depending on the value of `thing` at runtime.

This function would currently return `int | null`.

The coalesce needs to fold the types of all the potentially nullable fields together.

Reply:
Good catch!